### ISSUE TITLE
(main) unexpected error status OD_DETACH (4)

### ISSUE THREAD

**Issue by rostamiani** [yandex/odyssey#209]
I installed Odyssey on a different server other than PostgreSQL server and tested the throughput using Apache JMeter.
With high amount of connections, there is some errors. sometimes about 30% of responses!

This is the error message in the Odyssey logs:

```
(main) unexpected error status OD_DETACH (4)
```
I found no explanation about this error. just a commit here: https://github.com/yandex/odyssey/pull/131/commits/38fe99d5c140fc0592e9ab94bebe6c06785d5e2e

**Comment by x4m** [comment]
Hi! Thanks for showing this.
Seems like a bug, client disconnected right after finishing transaction. And there is not proper code that will handle this situation. Beside noisy logging there is one more inportant downside: server connection will be expelled from pool, which is not good...
We will fix this, thanks!

**Comment by rostamiani** [comment]
> Hi! Thanks for showing this.
> Seems like a bug, client disconnected right after finishing transaction. And there is not proper code that will handle this situation. Beside noisy logging there is one more important downside: server connection will be expelled from pool, which is not good...
> We will fix this, thanks!

Thanks
I see some other errors if it helps:

This happens sometimes between  OD_DETACH errors:
```
(reset) not synchronized, closing
```

And sometimes these:

```
(main) ERROR 42501 permission denied for sequence session_sid_seq
(main) ERROR 42501 permission denied for sequence url_errors_id_seq
(main) server disconnected (read/write error): Connection timed out, status OD_ESERVER_WRITE
(main) server disconnected (read/write error): Connection timed out, status OD_ESERVER_READ
```

**Comment by Gorilaz** [comment]
Hello, guys! I want to know the status of this issue. I hit it today on a production system.

**Comment by x4m** [comment]
@Gorilaz which version do you use? This issue should be fixed now in master. And we are planing 1.2 soon as long as online restart will be stable enough.

**Comment by Gorilaz** [comment]
@x4m 
console=> show version;
      version      
-------------------
 1-40fd90e-release

We installed odyssey in the next way:
git clone --depth 1 git://github.com/yandex/odyssey.git
And then we'll build the project like instructions says

**Comment by x4m** [comment]
40fd90e is a version where I'd expect the problem to be fixed...

What else besides logs with this message did you observe?

**Comment by Gorilaz** [comment]
Nothing suspicious was noticed. The load on database was average. 
97 2021-01-29T16:21:55Z info [none none] (stats) clients 36
97 2021-01-29T16:21:55Z info [none none] (stats) worker[0]: msg (1359430772 allocated, 0 cached, 1359470940 freed, 0 cache_size), coroutines (14 active, 0 cached), clients_processed: 8449
97 2021-01-29T16:21:55Z info [none none] (stats) worker[1]: msg (1333665685 allocated, 0 cached, 1333705857 freed, 0 cache_size), coroutines (13 active, 0 cached), clients_processed: 8449
97 2021-01-29T16:21:55Z info [none none] (stats) [tradingview.tv_tradingview_user] 36 clients, 6 active servers, 30 idle servers, 8008 transactions/sec (55 usec) 8045 queries/sec (50 usec) 5110797 in bytes/sec,
97 2021-01-29T16:21:55Z info [none none] (stats) worker[2]: msg (1333927571 allocated, 0 cached, 1333967745 freed, 0 cache_size), coroutines (12 active, 0 cached), clients_processed: 8448
97 2021-01-29T16:22:25Z info [none none] (stats) system worker: msg (120778 allocated, 0 cached, 49 freed, 0 cache_size), coroutines (4 active, 0 cached) startup errors 0
97 2021-01-29T16:22:25Z info [none none] (stats) clients 36
97 2021-01-29T16:22:25Z info [none none] (stats) worker[1]: msg (1334283532 allocated, 0 cached, 1334323705 freed, 0 cache_size), coroutines (13 active, 0 cached), clients_processed: 8449
97 2021-01-29T16:22:25Z info [none none] (stats) worker[0]: msg (1360058466 allocated, 0 cached, 1360098637 freed, 0 cache_size), coroutines (14 active, 0 cached), clients_processed: 8450
97 2021-01-29T16:22:25Z info [none none] (stats) worker[2]: msg (1334392816 allocated, 0 cached, 1334432993 freed, 0 cache_size), coroutines (12 active, 0 cached), clients_processed: 8449
97 2021-01-29T16:22:25Z info [none none] (stats) [tradingview.tv_tradingview_user] 36 clients, 6 active servers, 30 idle servers, 8274 transactions/sec (69 usec) 8312 queries/sec (63 usec) 5385713 in bytes/sec,
97 2021-01-29T16:22:55Z info [none none] (stats) system worker: msg (120781 allocated, 0 cached, 49 freed, 0 cache_size), coroutines (4 active, 0 cached) startup errors 0
97 2021-01-29T16:22:55Z info [none none] (stats) clients 36
97 2021-01-29T16:22:55Z info [none none] (stats) [tradingview.tv_tradingview_user] 36 clients, 3 active servers, 33 idle servers, 8527 transactions/sec (52 usec) 8563 queries/sec (46 usec) 4963299 in bytes/sec,
97 2021-01-29T16:22:55Z info [none none] (stats) worker[1]: msg (1335054975 allocated, 0 cached, 1335095149 freed, 0 cache_size), coroutines (13 active, 0 cached), clients_processed: 8449
97 2021-01-29T16:22:55Z info [none none] (stats) worker[0]: msg (1360579305 allocated, 0 cached, 1360619477 freed, 0 cache_size), coroutines (14 active, 0 cached), clients_processed: 8450
97 2021-01-29T16:22:55Z info [none none] (stats) worker[2]: msg (1334864582 allocated, 0 cached, 1334904760 freed, 0 cache_size), coroutines (12 active, 0 cached), clients_processed: 8449
97 2021-01-29T16:23:25Z info [none none] (stats) system worker: msg (120788 allocated, 0 cached, 49 freed, 0 cache_size), coroutines (4 active, 0 cached) startup errors 0
97 2021-01-29T16:23:25Z info [none none] (stats) clients 36
97 2021-01-29T16:23:25Z info [none none] (stats) worker[1]: msg (1335732281 allocated, 0 cached, 1335772458 freed, 0 cache_size), coroutines (13 active, 0 cached), clients_processed: 8450
97 2021-01-29T16:23:25Z info [none none] (stats) worker[2]: msg (1335536783 allocated, 0 cached, 1335576964 freed, 0 cache_size), coroutines (12 active, 0 cached), clients_processed: 8450
97 2021-01-29T16:23:25Z info [none none] (stats) [tradingview.tv_tradingview_user] 36 clients, 4 active servers, 32 idle servers, 8642 transactions/sec (57 usec) 8679 queries/sec (52 usec) 4660870 in bytes/sec,
97 2021-01-29T16:23:25Z info [none none] (stats) worker[0]: msg (1361018385 allocated, 0 cached, 1361058558 freed, 0 cache_size), coroutines (14 active, 0 cached), clients_processed: 8450
97 2021-01-29T16:23:55Z info [none none] (stats) system worker: msg (120791 allocated, 0 cached, 49 freed, 0 cache_size), coroutines (4 active, 0 cached) startup errors 0
97 2021-01-29T16:23:55Z info [none none] (stats) clients 36
97 2021-01-29T16:23:55Z info [none none] (stats) worker[0]: msg (1361401835 allocated, 0 cached, 1361442009 freed, 0 cache_size), coroutines (14 active, 0 cached), clients_processed: 8450
97 2021-01-29T16:23:55Z info [none none] (stats) worker[1]: msg (1336288228 allocated, 0 cached, 1336328406 freed, 0 cache_size), coroutines (13 active, 0 cached), clients_processed: 8450
97 2021-01-29T16:23:55Z info [none none] (stats) [tradingview.tv_tradingview_user] 36 clients, 4 active servers, 32 idle servers, 7049 transactions/sec (49 usec) 7084 queries/sec (44 usec) 4241026 in bytes/sec,
97 2021-01-29T16:23:55Z info [none none] (stats) worker[2]: msg (1336050570 allocated, 0 cached, 1336090752 freed, 0 cache_size), coroutines (12 active, 0 cached), clients_processed: 8450
97 2021-01-29T16:24:14Z error [c4b33e571e26b sfc50763e3357] (main) unexpected error status OD_DETACH (4)
97 2021-01-29T16:24:14Z error [c1c88826b1f1c s0c4ab517ca82] (main) unexpected error status OD_DETACH (4)
97 2021-01-29T16:24:14Z error [ceca72d6e1cb1 sdf1c881529f0] (main) unexpected error status OD_DETACH (4)
97 2021-01-29T16:24:14Z error [ca2d9b23dada2 s13b4c15e782f] (main) unexpected error status OD_DETACH (4)
97 2021-01-29T16:24:15Z error [cd53fc646c6fb s0cda644fe415] (main) unexpected error status OD_DETACH (4)

**Comment by x4m** [comment]
So, nothing wrong actually is happening, just these log messages?
As far as I understand, the pooler is complaining that the client disconnected after finishing transaction without sending a termination message. I think I'll add a less verbose handler for this in a future releasess.

**Comment by Gorilaz** [comment]
We found out that we had a skb buffers overload. During this time we watched TCP timeouts, Udp Errors All (UdpRcvbufErrors)
Doubled the following parameters :
net.core.rmem_max = 33554432
net.core.rmem_default = 425984
net.core.wmem_default = 425984
net.core.wmem_max = 33554432
Since then there have been no errors. So we continue to implement odyssey on our production systems.
And one more question) When we change pool_size and then do reload odyssey we hit the situation when user got :

```
server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
The connection to the server was lost. Attempting reset: Succeeded.
```
But only those users who were attached to specific database which pool_size changed. Is it a normal behavior ?
